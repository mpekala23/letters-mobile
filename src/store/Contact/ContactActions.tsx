import {
  Contact,
  SET_ADDING,
  SET_EXISTING,
  CLEAR_CONTACTS,
  ContactActionTypes,
} from "./ContactTypes";

export function setAdding(contact: Contact): ContactActionTypes {
  return {
    type: SET_ADDING,
    payload: contact,
  };
}

export function setExisting(contacts: Contact[]): ContactActionTypes {
  return {
    type: SET_EXISTING,
    payload: contacts,
  };
}

export function clearContacts() {
  return {
    type: CLEAR_CONTACTS,
    payload: null,
  };
}
