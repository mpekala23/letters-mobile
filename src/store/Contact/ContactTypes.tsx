import { Facility } from "types";

export const SET_ADDING = "contact/set_adding";
export const SET_EXISTING = "contact/set_existing";
export const CLEAR_CONTACTS = "contact/clear_contacts";

export interface Contact {
  id: number;
  state: string;
  firstName: string;
  lastName: string;
  inmateNumber: string;
  relationship: string;
  facility: Facility | null;
}

export interface ContactState {
  adding: Contact;
  existing: Contact[];
}

interface SetAddingAction {
  type: typeof SET_ADDING;
  payload: Contact;
}

interface SetExistingAction {
  type: typeof SET_EXISTING;
  payload: Contact[];
}

interface ClearContactsAction {
  type: typeof CLEAR_CONTACTS;
  payload: null;
}

export type ContactActionTypes =
  | SetAddingAction
  | SetExistingAction
  | ClearContactsAction;
