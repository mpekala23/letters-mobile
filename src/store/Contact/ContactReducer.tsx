import { ContactState, ContactActionTypes, SET_ADDING, SET_EXISTING } from './ContactTypes';

const initialState: ContactState = {
  adding: {
    state: '',
    first_name: '',
    last_name: '',
    inmate_number: '',
    relationship: '',
    facility: null,
  },
  existing: [],
};

export default function ContactReducer(
  state = initialState,
  action: ContactActionTypes
): ContactState {
  const currentState = { ...state };
  switch (action.type) {
    case SET_ADDING:
      currentState.adding = action.payload;
      return currentState;
    case SET_EXISTING:
      currentState.existing = action.payload;
      return currentState;
    default:
      return state;
  }
}
