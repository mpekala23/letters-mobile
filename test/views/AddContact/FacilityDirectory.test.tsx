import * as React from 'react';
import { FacilityDirectoryScreen } from '@views';
import { render, toJSON, fireEvent } from '@testing-library/react-native';
import { SET_ADDING } from '@store/Contact/ContactTypes';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { getFacilities } from '@api';

jest.mock('@api', () => ({
  getFacilities: jest.fn().mockReturnValue([
    {
      name: 'Yukon Kskokwim Correctional Center',
      type: 'State Prison',
      address: 'P.O. Box 400',
      city: 'Bethel',
      state: 'Alaska',
      postal: '99559',
    },
  ]),
}));

const mockStore = configureStore([]);

const setup = (facilityOverrides = {}, routeOverrides = {}) => {
  const navigation = {
    navigate: jest.fn(),
    addListener: jest.fn(),
    setParams: jest.fn(),
  };
  const route = {
    params: {
      newFacility: null,
      physState: '',
      ...routeOverrides,
    },
  };
  const contact = {
    firstName: '',
    lastName: '',
    inmateNumber: '',
    relationship: '',
    facility: null,
    ...facilityOverrides,
  };
  const initialState = {
    adding: contact,
    existing: [],
  };
  const store = mockStore({
    contact: initialState,
  });

  const StoreProvider = ({ children }: { children: JSX.Element }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return {
    navigation,
    store,
    ...render(
      <FacilityDirectoryScreen
        navigation={navigation}
        contactState={initialState}
        route={route}
      />,
      {
        wrapper: StoreProvider,
      }
    ),
  };
};

function sleep(ms: number) {
  jest.useRealTimers();
  return new Promise((resolve) => setTimeout(resolve, ms));
}

describe('Facility Directory Screen', () => {
  it('should match snapshot', () => {
    const { container } = setup();
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it('should have next button be disabled until facility selected', async () => {
    const { navigation, getByText } = setup();
    await sleep(1);
    const nextButton = getByText('Next');
    fireEvent.press(nextButton);
    expect(
      getByText('Next').parentNode.props.style[1].backgroundColor
    ).toBeDefined();
    expect(navigation.navigate).toHaveBeenCalledTimes(0);
    fireEvent.press(getByText('Yukon Kskokwim Correctional Center'));
    fireEvent.press(nextButton);
    expect(getByText('Next').parentNode.props.style[1]).toEqual({});
    expect(navigation.navigate).toHaveBeenCalledTimes(1);
  });

  it('should have the selected facility have a pink background', async () => {
    const { getByText } = setup();
    await sleep(1);
    const facility = getByText('Yukon Kskokwim Correctional Center').parentNode;
    expect(facility.props.style[1].backgroundColor).toBe('white');
    expect(facility.props.style[2]).toEqual({});
    fireEvent.press(facility);
    expect(facility.props.style[2].borderColor).toBe('#FF7171');
    fireEvent.press(facility);
    expect(facility.props.style[1].borderColor).toBe('white');
    expect(facility.props.style[2]).toEqual({});
  });

  it('should dispatch facility info to the redux store when the next button is pressed', async () => {
    const { store, getByText } = setup();
    await sleep(1);
    const nextButton = getByText('Next');
    fireEvent.press(nextButton);
    expect(store.getActions().length).toBe(0);
    fireEvent.press(getByText('Yukon Kskokwim Correctional Center'));
    fireEvent.press(nextButton);
    expect(store.getActions().length).toBe(1);
    expect(store.getActions()[0]).toEqual({
      type: SET_ADDING,
      payload: {
        facility: {
          address: 'P.O. Box 400',
          city: 'Bethel',
          name: 'Yukon Kskokwim Correctional Center',
          postal: '99559',
          state: 'Alaska',
          type: 'State Prison',
        },
        firstName: '',
        inmateNumber: '',
        lastName: '',
        relationship: '',
      },
    });
  });

  it('should navigate to the add manually screen when the add manually button is pressed', () => {
    const { navigation, getByText } = setup();
    fireEvent.press(getByText('Add Manually'));
    expect(navigation.navigate).toHaveBeenCalledWith('AddManually', {
      phyState: '',
    });
  });

  it('should navigate to the review contact screen when the next button is pressed', async () => {
    const { navigation, getByText } = setup();
    await sleep(1);
    const nextButton = getByText('Next');
    fireEvent.press(getByText('Yukon Kskokwim Correctional Center'));
    fireEvent.press(nextButton);
    expect(navigation.navigate).toHaveBeenCalledWith('ReviewContact');
  });

  it('should load initial facility value from the redux store', async () => {
    const { getByText } = setup({
      state: 'MN',
      firstName: 'First',
      lastName: 'Last',
      inmateNumber: '6',
      relationship: 'Sister',
      facility: {
        address: 'P.O. Box 400',
        city: 'Bethel',
        name: 'Yukon Kskokwim Correctional Center',
        postal: '99559',
        state: 'AK',
        type: 'State Prison',
      },
    });
    await sleep(1);
    const facility = getByText('Yukon Kskokwim Correctional Center').parentNode;
    expect(facility.props.style[1].backgroundColor).toBe('white');
  });

  it('should load initial facility from manual add', async () => {
    const { navigation } = setup(
      {
        state: 'MN',
        firstName: 'First',
        lastName: 'Last',
        inmateNumber: '6',
        relationship: 'Sister',
        facility: null,
      },
      {
        newFacility: {
          address: 'P.O. Box 400',
          city: 'Bethel',
          name: 'New Yukon Kskokwim Correctional Center',
          postal: '99559',
          state: 'AK',
          type: 'State Prison',
        },
      }
    );
    await sleep(1);
    expect(navigation.setParams).toHaveBeenCalledWith({ newFacility: null });
  });

  it('should show hint message when contact state is Pennsylvania', async () => {
    const { getByTestId } = setup(
      {},
      {
        newFacility: {},
        phyState: 'Pennsylvania',
      }
    );
    await sleep(1);
    expect(getByTestId('hintText')).toBeDefined();
  });
});
