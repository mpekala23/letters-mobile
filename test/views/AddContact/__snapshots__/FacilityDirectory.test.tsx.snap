// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facility Directory Screen should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TouchableOpacity
    activeOpacity={1}
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "white",
            "elevation": 5,
            "paddingBottom": 8,
            "paddingHorizontal": 16,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignSelf": "flex-start",
              "flexDirection": "row",
              "paddingTop": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontFamily": "Poppins-Bold",
                  "fontSize": 24,
                }
              }
            >
              Facility Directory
            </Text>
          </View>
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={22}
            bbWidth={22}
            fill="none"
            focusable={false}
            height={22}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "margin": 16,
                },
                Object {
                  "flex": 0,
                  "height": 22,
                  "width": 22,
                },
              ]
            }
            vbHeight={22}
            vbWidth={22}
            width={22}
            xml="<svg width=\\"22\\" height=\\"22\\" viewBox=\\"0 0 22 22\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<path d=\\"M18.5833 20.75V3.41667C18.5833 2.84203 18.3551 2.29093 17.9487 1.8846C17.5424 1.47827 16.9913 1.25 16.4167 1.25H5.58333C5.0087 1.25 4.4576 1.47827 4.05127 1.8846C3.64494 2.29093 3.41667 2.84203 3.41667 3.41667V20.75M18.5833 20.75H3.41667M18.5833 20.75H20.75M18.5833 20.75H13.1667M3.41667 20.75H1.25M3.41667 20.75H8.83333M13.1667 20.75V15.3333C13.1667 15.046 13.0525 14.7705 12.8494 14.5673C12.6462 14.3641 12.3707 14.25 12.0833 14.25H9.91667C9.62935 14.25 9.3538 14.3641 9.15063 14.5673C8.94747 14.7705 8.83333 15.046 8.83333 15.3333V20.75M13.1667 20.75H8.83333M7.75 5.58333H8.83333M7.75 9.91667H8.83333M13.1667 5.58333H14.25M13.1667 9.91667H14.25\\" stroke=\\"black\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/>
</svg>
"
            xmlns="http://www.w3.org/2000/svg"
          >
            <RNSVGGroup
              fill={null}
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                d="M18.5833 20.75V3.41667C18.5833 2.84203 18.3551 2.29093 17.9487 1.8846C17.5424 1.47827 16.9913 1.25 16.4167 1.25H5.58333C5.0087 1.25 4.4576 1.47827 4.05127 1.8846C3.64494 2.29093 3.41667 2.84203 3.41667 3.41667V20.75M18.5833 20.75H3.41667M18.5833 20.75H20.75M18.5833 20.75H13.1667M3.41667 20.75H1.25M3.41667 20.75H8.83333M13.1667 20.75V15.3333C13.1667 15.046 13.0525 14.7705 12.8494 14.5673C12.6462 14.3641 12.3707 14.25 12.0833 14.25H9.91667C9.62935 14.25 9.3538 14.3641 9.15063 14.5673C8.94747 14.7705 8.83333 15.046 8.83333 15.3333V20.75M13.1667 20.75H8.83333M7.75 5.58333H8.83333M7.75 9.91667H8.83333M13.1667 5.58333H14.25M13.1667 9.91667H14.25"
                propList={
                  Array [
                    "stroke",
                    "strokeWidth",
                    "strokeLinecap",
                    "strokeLinejoin",
                  ]
                }
                stroke={4278190080}
                strokeLinecap={1}
                strokeLinejoin={1}
                strokeWidth={2}
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
        <View
          pointerEvents="auto"
          style={
            Object {
              "backgroundColor": "#F2F2F2",
              "borderRadius": 8,
              "height": 50,
              "marginBottom": 10,
              "marginTop": 4,
              "overflow": "hidden",
              "width": "100%",
            }
          }
          testID="parent"
        >
          <View
            testID="invalid"
          />
          <View
            testID="unfocused"
          />
          <TextInput
            allowFontScaling={true}
            multiline={false}
            numberOfLines={1}
            placeholder="Search Facility"
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Array [
                Object {
                  "borderColor": "rgba(0,0,0,0)",
                  "borderRadius": 8,
                  "borderWidth": 2,
                  "color": "black",
                  "fontSize": 16,
                  "paddingHorizontal": 10,
                  "textAlignVertical": "center",
                  "width": "100%",
                },
                Object {
                  "color": "#282828",
                },
                Object {},
                Object {
                  "height": 50,
                },
                Object {
                  "borderRadius": 20,
                },
                Object {
                  "fontFamily": "Poppins-Regular",
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "height": 0,
                "width": "100%",
                "zIndex": 5,
              }
            }
          >
            <ScrollView
              keyboardShouldPersistTaps="always"
              nestedScrollEnabled={true}
              scrollEnabled={true}
              style={
                Object {
                  "flex": 1,
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </View>
      <ScrollView
        collapsable={false}
        contentContainerStyle={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
        overScrollMode="always"
        refreshControl={
          <RefreshControl
            onRefresh={[Function]}
            refreshing={true}
          />
        }
        scrollEnabled={true}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "height": 24,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "marginTop": 20,
            }
          }
        >
          <View />
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 16,
                },
                Object {
                  "marginBottom": 20,
                },
              ]
            }
          >
            Don't see the facility you're looking for?
          </Text>
          <TouchableOpacity
            activeOpacity={0.7}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#FF7171",
                  "borderRadius": 16,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 5,
                  "paddingHorizontal": 14,
                },
                Object {},
                Object {
                  "width": 200,
                },
                Object {},
                Object {
                  "elevation": 4,
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.22,
                },
              ]
            }
            testID="clickable"
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "Poppins-Medium",
                  },
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "fontWeight": "700",
                    "padding": 6,
                  },
                  Object {},
                  Object {},
                  Object {},
                ]
              }
            >
              Add Manually
            </Text>
          </TouchableOpacity>
        </View>
        <View
          style={
            Object {
              "height": 24,
              "width": "100%",
            }
          }
        />
      </ScrollView>
    </View>
  </TouchableOpacity>
</View>
`;
